### PR TITLE
TVPF-16399 Implement ticket-based authentication in bonefish router

### PR DESCRIPTION
**Pull request by oleh-synelnykov**

### CODE REVIEW

File: src/bonefish/messages/wamp_authenticate_extra.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <bonefish/messages/wamp_authenticate_extra.hpp>
+
+#include <stdexcept>
```
Comment (pavlodenysov-vizio) [inline src/bonefish/messages/wamp_authenticate_extra.cpp]:
```suggestion
```

File: src/bonefish/messages/wamp_authenticate_extra.hpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+#include <string>
+#include <map>
+
+namespace bonefish {
+
+class wamp_authenticate_extra
+{
+public:
+    wamp_authenticate_extra();
+    virtual ~wamp_authenticate_extra();
```
Comment (pavlodenysov-vizio) [inline src/bonefish/messages/wamp_authenticate_extra.hpp]:
```suggestion
    ~wamp_authenticate_extra();
```

File: src/bonefish/messages/wamp_challenge_extra.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <bonefish/messages/wamp_challenge_extra.hpp>
+
+#include <stdexcept>
```
Comment (pavlodenysov-vizio) [inline src/bonefish/messages/wamp_challenge_extra.cpp]:
```suggestion
```

File: src/bonefish/messages/wamp_challenge_extra.hpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+#include <string>
+#include <map>
+
+namespace bonefish {
+
+class wamp_challenge_extra
+{
+public:
+    wamp_challenge_extra();
+    virtual ~wamp_challenge_extra();
```
Comment (pavlodenysov-vizio) [inline src/bonefish/messages/wamp_challenge_extra.hpp]:
```suggestion
    ~wamp_challenge_extra();
```

File: src/bonefish/messages/wamp_hello_details_advanced.hpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+#include <string>
+#include <vector>
+
+namespace bonefish {
+
+class wamp_hello_details_advanced
+{
+public:
+    wamp_hello_details_advanced();
+    virtual ~wamp_hello_details_advanced();
```
Comment (pavlodenysov-vizio) [inline src/bonefish/messages/wamp_hello_details_advanced.hpp]:
```suggestion
    ~wamp_hello_details_advanced();
```

File: src/bonefish/router/wamp_router_impl.cpp [hunk 6]
```diff
@@ -179,19 +189,70 @@ void wamp_router_impl::process_hello_message(const wamp_session_id& session_id,
...
+        // check if any of the requested authentication methods
+        // is supported by provided authenticator
+        std::string selected_auth_method;
+        const std::vector<std::string>& supported = m_authenticator->methods();
+        for (const auto& method : hello_details.get_auth_methods()) {
+            if (auto i = std::find(supported.begin(), supported.end(), method); i != supported.end()) {
+                selected_auth_method = *i;
+                break;
+            }
+        }
```
Comment (pavlodenysov-vizio) [inline src/bonefish/router/wamp_router_impl.cpp]:
```cpp
find_if(hello_details.get_auth_methods(), method_is_supported);
```

File: src/bonefish/router/wamp_router_impl.cpp [hunk 7]
```diff
@@ -179,19 +189,70 @@ void wamp_router_impl::process_hello_message(const wamp_session_id& session_id,
...
+        welcome_message->set_session_id(session_id);
+        welcome_message->set_details(m_welcome_details.marshal(welcome_message->get_zone()));
+
+        // If we fail to send the welcome message it is most likely that the
+        // underlying network connection has been closed/lost which means
+        // that the component is no longer reachable on this session. So all
+        // we do here is trace the fact that this event occured.
+        BONEFISH_TRACE("%1%, %2%", *session % *welcome_message);
+        if (!session->get_transport()->send_message(std::move(*welcome_message))) {
+            BONEFISH_TRACE("failed to send the welcome message: network failure");
```
Comment (pavlodenysov-vizio) [inline src/bonefish/router/wamp_router_impl.cpp]: `BONEFISH_ERROR` is changed to `BONEFISH_TRACE`, is it intended?

Also, in above cases errors are logged by `BONEFISH_TRACE`.

File: src/bonefish/router/wamp_router_impl.cpp [hunk 8]
```diff
@@ -220,6 +281,78 @@ void wamp_router_impl::process_goodbye_message(const wamp_session_id& session_id
...
+{
+    auto session_itr = m_sessions.find(session_id);
+    if (session_itr == m_sessions.end()) {
+        throw std::logic_error("session does not exist");
+    }
+
+    auto session_auth_itr = m_session_auth_info.find(session_id);
+    if (session_auth_itr == m_session_auth_info.end()) {
+        throw std::logic_error("session does not have auth info");
+    }
```
Comment (pavlodenysov-vizio) [inline src/bonefish/router/wamp_router_impl.cpp]: protocol violation

File: src/bonefish/router/wamp_router_impl.cpp (hunk 6 above)
Comment (oleh-synelnykov) [reply src/bonefish/router/wamp_router_impl.cpp]: done

File: src/bonefish/messages/wamp_authenticate_extra.cpp (hunk 1 above)
Comment (oleh-synelnykov) [reply src/bonefish/messages/wamp_authenticate_extra.cpp]: done

File: src/bonefish/messages/wamp_authenticate_extra.hpp (hunk 2 above)
Comment (oleh-synelnykov) [reply src/bonefish/messages/wamp_authenticate_extra.hpp]: done

File: src/bonefish/messages/wamp_challenge_extra.cpp (hunk 3 above)
Comment (oleh-synelnykov) [reply src/bonefish/messages/wamp_challenge_extra.cpp]: done

File: src/bonefish/messages/wamp_challenge_extra.hpp (hunk 4 above)
Comment (oleh-synelnykov) [reply src/bonefish/messages/wamp_challenge_extra.hpp]: done

File: src/bonefish/messages/wamp_hello_details_advanced.hpp (hunk 5 above)
Comment (oleh-synelnykov) [reply src/bonefish/messages/wamp_hello_details_advanced.hpp]: done

File: src/bonefish/router/wamp_router_impl.cpp (hunk 7 above)
Comment (oleh-synelnykov) [reply src/bonefish/router/wamp_router_impl.cpp]: done

File: src/bonefish/router/wamp_router_impl.cpp (hunk 8 above)
Comment (oleh-synelnykov) [reply src/bonefish/router/wamp_router_impl.cpp]: done

**Pull request merged**